integrationTests/testNetwork.go

```diff
@@ -0,0 +1,407 @@
+	for i, wallet := range net.Wallets {
+		account := net.GetAccountHandler(wallet.Address)
+		require.Equal(net.T, wallet.Nonce, account.GetNonce(),
+			fmt.Sprintf("wallet %d has nonce out of sync", i))
+	}
+}
+
+// TODO cannot define SCQueryInt() because it requires vm.GetIntValueFromSC()
+// which causes an import cycle.
+// func (net *TestNetwork) SCQueryInt(contract Address, function string, args ...[]byte) *big.Int {
+// 	shardID := net.ShardOfAddress(contract)
+// 	firstNodeInShard := net.NodesSharded[shardID][0]
+
+// 	return vm.GetIntValueFromSC(
+// 		net.DefaultGasSchedule,
+// 		firstNodeInShard.AccntState,
+// 		contract,
+// 		function,
+// 		args...)
+// }
+
+func (net *TestNetwork) createAdvertiser() {
+	net.Advertiser = CreateMessengerWithKadDht(net.AdvertiserAddress)
+	err := net.Advertiser.Bootstrap(0)
+	net.handleOrBypassError(err)
+}
+
+func (net *TestNetwork) createNodes() {
+	net.Nodes = CreateNodes(
+		net.NumShards,
+		net.NodesPerShard,
+		net.NodesInMetashard,
+		GetConnectableAddress(net.Advertiser))
+}
+
+func (net *TestNetwork) indexProposers() {
+	net.Proposers = make([]int, net.NumShards+1)
+	for i := 0; i < net.NumShards; i++ {
+		net.Proposers[i] = i * net.NodesPerShard
+	}
+	net.Proposers[net.NumShards] = net.NumShards * net.NodesPerShard
+}
+
+func (net *TestNetwork) mapNodesByShard() {
+	net.NodesSharded = make(NodesByShardMap)
+	for _, node := range net.Nodes {
+		shardID := node.ShardCoordinator.SelfId()
+		_, exists := net.NodesSharded[shardID]
+		if !exists {
+			net.NodesSharded[shardID] = make(NodeSlice, 0)
+		}
+		net.NodesSharded[shardID] = append(net.NodesSharded[shardID], node)
+	}
+}
+
+func (net *TestNetwork) startNodes() {
+	DisplayAndStartNodes(net.Nodes)
+}
+
+func (net *TestNetwork) initDefaults() {
+	net.DeploymentAddress = make(Address, 32)
+	net.DefaultNode = net.Nodes[0]
+	net.DefaultGasPrice = MinTxGasPrice
+	net.DefaultGasSchedule = nil
+	net.DefaultVM = factory.ArwenVirtualMachine
+
+	defaultNodeShardID := net.DefaultNode.ShardCoordinator.SelfId()
+	net.MinGasLimit = MinTxGasLimit
+	net.MaxGasLimit = net.DefaultNode.EconomicsData.MaxGasLimitPerBlock(defaultNodeShardID) - 1
+}
+
+func (net *TestNetwork) closeAdvertiser() {
+	err := net.Advertiser.Close()
+	net.handleOrBypassError(err)
+}
+
+func (net *TestNetwork) closeNodes() {
+	for _, node := range net.Nodes {
+		err := node.Messenger.Close()
+		net.handleOrBypassError(err)
+	}
+}
+
+func (net *TestNetwork) firstNodeInShard(shardID ShardIdentifier) *TestProcessorNode {
+	firstNodeInShard := net.NodesSharded[shardID][0]
+	require.NotNil(net.T, firstNodeInShard)
+	return firstNodeInShard
+}
+
+func (net *TestNetwork) firstNodeInShardOfAddress(address Address) *TestProcessorNode {
+	shardID := net.ShardOfAddress(address)
+	firstNodeInShard := net.NodesSharded[shardID][0]
+	require.NotNil(net.T, firstNodeInShard)
+	return firstNodeInShard
+}
+
+func (net *TestNetwork) handleOrBypassError(err error) {
+	if net.BypassErrorsOnce == true {
```

Comment:
if net.BypassErrorsOnce {

Reply:
Remnant of a previous condition, removed.